runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,127 @@
+local keymap = {}
+
+--- Add a new keymap.
+--- Examples:
+--- <pre>
+---   -- You can make keymaps worth lua functions
+---   vim.keymap.set('n', 'lhs', function() print("real lua function") end)
+---
+---   -- Can use it to map multiple modes
+---   vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { silent = true, buffer=true })
+---
+---   -- You can map a key for specific buffer
+---   vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", { silent = true, buffer = 5 })
+---
+---   -- Also expr keymaps too. Notice the replace_termcodes when making lua expr maps.
+---   vim.keymap.set('i', '<Tab>', function()
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { silent = true, expr = true, replace_termcodes = true })
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
+--@param mode (string|table):   Same mode short names as |nvim_set_keymap|.
+---                             Can also be list of modes to create keymap on multiple modes.
+--@param lhs (string):          Left-hand-side |{lhs}| of the mapping.
+--@param rhs (string|function): Right-hand-side |{rhs}| of the mapping. Can also be a lua function.
+--
+--@param opts (table): A table of |:map-arguments| such as "silent". See |nvim_set_keymap()|
+---       along with "replace_termcodes". When replace_termcodes is true with
+---       lua function expr maps |nvim_replace_termcodes| is applied to return
+---       value of mapped function.
+---       One difference compared to |nvim_set_keymap| is here "remap" used as option
+---       instead of |noremap| . It's meaning is opposite of |noremap| and it's false by default.
+--@see |nvim_set_keymap|
+---
+function keymap.set(modes, lhs, rhs, opts)
+  opts = opts or {}
+
+  vim.validate {
+    mode = {modes, {'s', 't'}},
+    lhs = {lhs, 's'},
+    rhs = {rhs, {'s', 'f'}},
+    opts = {opts, 't'}
+  }
```

Comment:
Putting `vim.validate` at the very top of the function is a bit more idiomatic.

```suggestion
  vim.validate {
    mode = {modes, {'s', 't'}},
    lhs = {lhs, 's'},
    rhs = {rhs, {'s', 'f'}},
    opts = {opts, 't', true}
  }
  
  opts = opts or {}
```